File: CPP/CASTest.cpp
#include<iostream>
#include<atomic>
#include<thread>
using namespace std;

class CAS
{
private:
    std::atomic<bool> flag;
public:
    CAS():flag(false) {}
    ~CAS() = default;
    CAS(const CAS& s) = delete;
    CAS& operator=(const CAS&) = delete;
    void lock()
    {
        bool expect = false;
        while (!flag.compare_exchange_strong(expect, true))
        {
            expect = false;
        }
    }
    void unlock()
    {
        flag.store(false);
    }

};

int num = 0;
CAS sm;

void fun()
{
    for (int i = 0;i < 1000;i++)
    {
        sm.lock();
        cout << "num:" << ++num << endl;
        sm.unlock();
    }
}

int main()
{
    thread t1(fun);
    thread t2(fun);
    t1.join();
    t2.join();
    return 0;
}

File: JAVA/CASTest.java
package practise;

import java.util.concurrent.atomic.AtomicReference;

class Spinlock {
    private final AtomicReference<Thread> cas;

    Spinlock(AtomicReference<Thread> cas) {
        this.cas = cas;
    }

    public void lock() {
        Thread current = Thread.currentThread();
        // 利用CAS
        while (!cas.compareAndSet(null, current)) {
            // DO nothing
            System.out.println("I am spinning");
        }
    }

    public void unlock() {
        Thread current = Thread.currentThread();
        cas.compareAndSet(current, null);
    }
}

class Task implements Runnable {
    private final AtomicReference<Thread> cas;
    private final Spinlock slock;

    Task(AtomicReference<Thread> cas) {
        this.cas = cas;
        this.slock = new Spinlock(cas);
    }

    @Override
    public void run() {
        slock.lock(); //上锁
        for (int i = 0; i < 10; i++) {
            System.out.println(i);
        }
        slock.unlock();
    }
}

public class CASTest {
    public static void main(String[] args) {
        AtomicReference<Thread> cas = new AtomicReference<>();
        Thread thread1 = new Thread(new Task(cas));
        Thread thread2 = new Thread(new Task(cas));
        thread1.start();
        thread2.start();
    }
}
